### DISCUSSION TITLE
WidgetRef & Ref

### DISCUSSION THREAD

**Discussion by felipe-xavier** [rrousselGit/riverpod#1872] [General]
After upgrading to riverpod 2 there is no more Reader typedef and I noticed we have two ways to reference a ref object, using WidgetRef and Ref. What are the differences and why both?

**Comment by mellowcello77** [comment]
Gosh I am also looking for Reader, or a workaround desperately. 

I have passed "Reader reader" around between all my providers and classes, it works so well.

`final accountsViewSvc = ChangeNotifierProvider.autoDispose<AccountsViewSvc>((ref) => AccountsViewSvc(read: ref.read));`

```
class AccountsViewSvc extends ChangeNotifier {
  AccountsViewSvc({required this.read});
  final Reader read;

}
```


Hope anyone can provide some info on this or how this is expected to work in 2.0

**Reply by mellowcello77** [reply]
I don't know if this answers only my question about Reader being deprecated  - https://github.com/rrousselGit/riverpod/discussions/1639

**Reply by felipe-xavier** [reply]
Hey @mellowcello77 , in this case you will really need to replace, code wise, your `Reader read` by `Ref ref`, and then call `ref.read()`.

Would be like this:

```Dart
final accountsViewSvc = ChangeNotifierProvider.autoDispose<AccountsViewSvc>((ref) => AccountsViewSvc(ref: ref));

class AccountsViewSvc extends ChangeNotifier {
  AccountsViewSvc({required this.ref});
  final Ref ref;
}
```

Unfortunately it will require refactoring code wise.

**Reply by mellowcello77** [reply]
Thank you so much, this worked for me. I replaced everything with Ref ref and ref.read() as you suggested - happy days

**Comment by rrousselGit** [comment]
Using Ref is always the best option.

Using it means you moved the logic to a provider. So that's great

Try avoiding passing WidgetRef around

**Reply by janosgy** [reply]
hi @rrousselGit 
I understand the reasons, and I've updated my code to use Ref instead of Reader, but I'm not sure about my tests.
I guess this doc is outdated, [How to test an object that receives read as a parameter of its constructor?](https://docs-v2.riverpod.dev/docs/concepts/combining_providers#how-to-test-an-object-that-receives-read-as-a-parameter-of-its-constructor) 
but what's the new approach?

A simple replacement:
`repository = Repository(container);`
Results in `The argument type 'ProviderContainer' can't be assigned to the parameter type 'Ref<Object?>'`

**Reply by janosgy** [reply]
My bad. I had to read the guide carefully, the title was a bit misleading "object that receives read as a parameter" (because it receives Ref now).
But I realized that I should get my repository through its provider.

**Reply by rrousselGit** [reply]
The doc linked if out of date. I'll update it
